File: 1351.count-negative-numbers-in-a-sorted-matrix.cpp
/*
 * @lc app=leetcode id=1351 lang=cpp
 *
 * [1351] Count Negative Numbers in a Sorted Matrix
 */

#include <bits/stdc++.h>
#include "lassert/lassert.h"

using namespace std;

// @lc code=start
class Solution {
 public:
  int countNegatives(vector<vector<int>>& grid) {
    int ret = 0;
    for (const auto& row : grid) {
      for (auto it = row.rbegin(); it != row.rend(); it++) {
        if (*it >= 0) {
          break;
        }
        ret++;
      }
    }
    return ret;
  }
};
// @lc code=end

int main() {
  Solution s;
  vector<vector<int>> grid;
  grid = {{4, 3, 2, -1}, {3, 2, 1, -1}, {1, 1, -1, -2}, {-1, -1, -2, -3}};
  ASSERT_EQ(s.countNegatives(grid), 8);
  grid = {{3, 2}, {1, 0}};
  ASSERT_EQ(s.countNegatives(grid), 0);
  grid = {{1, -1}, {-1, -1}};
  ASSERT_EQ(s.countNegatives(grid), 3);
  grid = {{-1}};
  ASSERT_EQ(s.countNegatives(grid), 1);
  return 0;
}

File: 1351.count-negative-numbers-in-a-sorted-matrix.rs
/*
 * @lc app=leetcode id=1351 lang=rust
 *
 * [1351] Count Negative Numbers in a Sorted Matrix
 */

struct Solution;

#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode { next: None, val }
    }
}

#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<std::rc::Rc<std::cell::RefCell<TreeNode>>>,
    pub right: Option<std::rc::Rc<std::cell::RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

#[macro_export]
macro_rules! list_node {
    () => {
        None
    };
    ( $x:expr,$( $y:expr ),* ) => {
        {
            let mut head = Some(Box::new(ListNode::new($x)));
            let mut tail = &mut head;
            $(
                let node = Some(Box::new(ListNode::new($y)));
                tail.as_mut().unwrap().next = node;
                tail = &mut tail.as_mut().unwrap().next;
            )*
            head
        }
    };
}

// @lc code=start
impl Solution {
    pub fn count_negatives(grid: Vec<Vec<i32>>) -> i32 {
        let mut ret = 0;
        for row in grid {
            for n in row {
                if n < 0 {
                    ret += 1;
                }
            }
        }
        ret
    }
}
// @lc code=end

#[test]
fn test1() {
    assert_eq!(
        Solution::count_negatives(vec![
            vec![4, 3, 2, -1],
            vec![3, 2, 1, -1],
            vec![1, 1, -1, -2],
            vec![-1, -1, -2, -3]
        ]),
        8
    );
}

#[test]
fn test2() {
    assert_eq!(Solution::count_negatives(vec![vec![3, 2], vec![1, 0]]), 0);
}
